File: sdks/rust/examples/quickstart.rs
/// Quickstart example: minimal end-to-end O2 Exchange flow.
///
/// Demonstrates: generate wallet, setup account, create session, place order,
/// check order status, cancel order.
use o2_sdk::{crypto, Network, O2Client, OrderType, Side};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let mut client = O2Client::new(Network::Testnet);

    // 1. Generate a wallet
    let wallet = client.generate_wallet()?;
    println!(
        "Owner address: {}",
        crypto::to_hex_string(&wallet.b256_address)
    );

    // 2. Setup account (creates account, mints via faucet, whitelists)
    println!("Setting up account...");
    let account = client.setup_account(&wallet).await?;
    let trade_account_id = account.trade_account_id.unwrap();
    println!("Trade account: {trade_account_id}");

    // 3. Fetch available markets
    let markets = client.get_markets().await?;
    println!("Available markets:");
    for m in &markets {
        println!("  {} ({})", m.symbol_pair(), m.market_id);
    }

    let market = markets[0].clone();
    let market_pair = market.symbol_pair();
    println!("\nTrading on: {market_pair}");

    // 4. Create a trading session
    println!("Creating session...");
    let mut session = client
        .create_session(
            &wallet,
            &[&market_pair],
            std::time::Duration::from_secs(30 * 24 * 3600),
        )
        .await?;
    println!("Session created, expiry: {}", session.expiry);

    // 5. Place a spot buy order (low price, unlikely to fill)
    println!("Placing buy order...");
    let price = market.price("0.001")?;
    let quantity = market.quantity("100")?;
    let result = client
        .create_order(
            &mut session,
            &market,
            Side::Buy,
            price,
            quantity,
            OrderType::Spot,
            true, // settle first
            true, // collect orders
        )
        .await;

    match result {
        Ok(resp) => {
            if resp.is_success() {
                println!(
                    "Order placed! tx_id: {}",
                    resp.tx_id.as_deref().unwrap_or("?")
                );
                if let Some(orders) = &resp.orders {
                    for order in orders {
                        println!("  Order ID: {}", order.order_id);
                        println!("  Side: {}", order.side);
                        println!("  Price: {}", order.price);
                        println!("  Quantity: {}", order.quantity);

                        // 6. Cancel the order
                        let oid = &order.order_id;
                        println!("\nCancelling order {oid}...");
                        let cancel = client.cancel_order(&mut session, oid, &market_pair).await;
                        match cancel {
                            Ok(_) => println!("Order cancelled successfully."),
                            Err(e) => println!("Cancel failed: {e}"),
                        }
                    }
                }
            } else {
                println!("Order failed: {:?}", resp.message);
            }
        }
        Err(e) => {
            println!("Order error: {e}");
            println!("(This is expected if account has insufficient balance.)");
        }
    }

    // 7. Check balances
    println!("\nChecking balances...");
    let balances = client.get_balances(&trade_account_id).await?;
    for (symbol, bal) in &balances {
        println!(
            "  {}: available={}, locked={}, unlocked={}",
            symbol, bal.trading_account_balance, bal.total_locked, bal.total_unlocked,
        );
    }

    println!("\nQuickstart complete!");
    Ok(())
}

File: sdks/python/examples/quickstart.py
"""Minimal end-to-end flow: generate wallet, setup account, place/cancel orders."""

import asyncio

from o2_sdk import Network, O2Client, OrderSide, OrderType


async def main():
    client = O2Client(network=Network.TESTNET)

    # 1. Generate a wallet
    owner = client.generate_wallet()
    print(f"Owner address: {owner.b256_address}")

    # 2. Setup account (creates account, funds via faucet, whitelists)
    account = await client.setup_account(owner)
    print(f"Trade account ID: {account.trade_account_id}")

    # 3. Get available markets
    markets = await client.get_markets()
    market = markets[0]
    print(f"Trading on: {market.pair}")

    # 4. Wait for faucet funds to arrive (may need multiple mints)
    print("Waiting for faucet funds...")
    await asyncio.sleep(5)

    # Optionally mint again to get more funds
    try:
        await asyncio.sleep(60)  # Faucet cooldown
        await client.api.mint_to_contract(account.trade_account_id)
        await asyncio.sleep(5)
    except Exception as e:
        print(f"Second mint: {e}")

    # 5. Create a trading session
    session = await client.create_session(
        owner=owner,
        markets=[market.pair],
        expiry_days=1,
    )
    print(f"Session created, expires: {session.session_expiry}")

    # 6. Place a spot buy order using high-level create_order
    depth = await client.get_depth(market.pair, precision=10)
    price = market.format_price(int(depth.best_ask.price)) * 0.5 if depth.best_ask else 0.01

    quantity = max(
        int(market.min_order) / market.scale_price(price) * 1.1,
        1.0,
    )

    print(f"Placing buy order: price={price}, quantity={quantity}")
    result = await client.create_order(
        session=session,
        market=market.pair,
        side=OrderSide.BUY,
        price=price,
        quantity=quantity,
        order_type=OrderType.SPOT,
        settle_first=True,
        collect_orders=True,
    )

    if result.success:
        print(f"Order placed! tx_id: {result.tx_id}")
        if result.orders:
            order = result.orders[0]
            print(f"Order ID: {order.order_id}")

            # 7. Check order status
            orders = await client.get_orders(account.trade_account_id, market.pair, is_open=True)
            print(f"Open orders: {len(orders)}")

            # 8. Build and submit a typed batch with the fluent builder
            batch = (
                client.actions_for(market)
                .settle_balance()
                .cancel_order(order.order_id)
                .create_order(OrderSide.BUY, "0.01", "1", OrderType.POST_ONLY)
                .build()
            )
            batch_result = await client.batch_actions(actions=[batch], session=session)
            print(f"Batch submitted: {batch_result.success}")
    else:
        print(f"Order failed: {result.message}")

    # 9. Check balances
    balances = await client.get_balances(account.trade_account_id)
    for symbol, balance in balances.items():
        print(f"  {symbol}: available={balance.trading_account_balance}")

    await client.close()
    print("Done!")


if __name__ == "__main__":
    asyncio.run(main())
